mobile_verifier/src/boosting_oracles/data_sets.rs

```diff
@@ -0,0 +1,632 @@
+            db::insert_new_data_set(pool, &new_data_set.key, Self::TYPE, new_data_set.timestamp)
+                .await?;
+        }
+        Ok(())
+    }
+
+    async fn fetch_next_available_data_set(
+        &mut self,
+        store: &FileStore,
+        pool: &PgPool,
+        data_set_directory: &Path,
+    ) -> anyhow::Result<Option<NewDataSet>> {
+        self.check_for_available_data_sets(store, pool).await?;
+
+        tracing::info!("Checking for new {} data sets", Self::TYPE.to_prefix());
+        let latest_unprocessed_data_set =
+            db::fetch_latest_unprocessed_data_set(pool, Self::TYPE, self.timestamp()).await?;
+
+        let Some(latest_unprocessed_data_set) = latest_unprocessed_data_set else {
+            return Ok(None);
+        };
+
+        let path = get_data_set_path(
+            data_set_directory,
+            Self::TYPE,
+            latest_unprocessed_data_set.time_to_use,
+        );
+
+        if !latest_unprocessed_data_set.status.is_downloaded() {
+            download_data_set(store, &latest_unprocessed_data_set.filename, &path).await?;
+            latest_unprocessed_data_set.mark_as_downloaded(pool).await?;
+            tracing::info!(
+                data_set = latest_unprocessed_data_set.filename,
+                "Data set download complete"
+            );
+        }
+
+        self.update(Path::new(&path), latest_unprocessed_data_set.time_to_use)?;
+
+        Ok(Some(latest_unprocessed_data_set))
+    }
+}
+
+pub struct DataSetDownloaderDaemon<A, B, C> {
+    pool: PgPool,
+    data_sets: HexBoostData<A, B, C>,
+    store: FileStore,
+    oracle_boosting_sink: FileSinkClient,
+    data_set_directory: PathBuf,
+    new_coverage_object_signal: Receiver<()>,
+}
+
+#[derive(FromRow)]
+pub struct NewDataSet {
+    filename: String,
+    time_to_use: DateTime<Utc>,
+    status: DataSetStatus,
+}
+
+impl NewDataSet {
+    async fn mark_as_downloaded(&self, pool: &PgPool) -> anyhow::Result<()> {
+        db::set_data_set_status(pool, &self.filename, DataSetStatus::Downloaded).await?;
+        Ok(())
+    }
+
+    async fn mark_as_processed(&self, pool: &PgPool) -> anyhow::Result<()> {
+        db::set_data_set_status(pool, &self.filename, DataSetStatus::Processed).await?;
+        Ok(())
+    }
+}
+
+#[derive(Copy, Clone, sqlx::Type)]
+#[sqlx(type_name = "data_set_status")]
+#[sqlx(rename_all = "lowercase")]
+pub enum DataSetStatus {
+    Pending,
+    Downloaded,
+    Processed,
+}
+
+impl DataSetStatus {
+    pub fn is_downloaded(&self) -> bool {
+        matches!(self, Self::Downloaded)
+    }
+}
+
+impl<A, B, C> ManagedTask for DataSetDownloaderDaemon<A, B, C>
+where
+    A: DataSet,
+    B: DataSet,
+    C: DataSet,
+{
+    fn start_task(
+        self: Box<Self>,
+        shutdown: triggered::Listener,
+    ) -> futures::prelude::future::LocalBoxFuture<'static, anyhow::Result<()>> {
+        let handle = tokio::spawn(async move {
+            #[rustfmt::skip]
```

Comment:
Is there a reason to skip all these `tokio::select!`s?

Reply:
rustfmt butchers them. it's appalling. It might be fixed at this point tho